architecture/NFT.md

```diff
@@ -0,0 +1,214 @@
+a token, if there is some privledged judicial body.
+
+But each species may have it's own special actions.
+A human-name token with a fixed name may have a special
+action `setPubkey` to update the public key associated
+with the name NFT. A BNS token may have a few like
+`setTrustedNodes`, `setRecentHeader`, `setTxCodec`, etc.
+These will all update the species-specific payload.
+
+The owner may not want to worry about updating the blockchain
+nodes all the time. Especially when the owner is a 8 of 10
+multisig contract for security. Or a DAO governance system.
+Thus, we also want a way for them to delegate approval to
+other actors for specific tasks. This should be fine-grained
+to prevent abuse.
+
+Of course, adding this concept of `approvals` means there is
+now another generic action common across all species...
+`grantApproval`.
+
+## Interfaces
+
+`id: []byte`: Every species will have it's own bucket and it's
+own primary key, which will be unique. They will generally want
+to add some validation logic as to what value ids are (such
+as ascii alphanumeric).
+
+`owner: Address`: Every token has one owner. This owner has full
+control of the token, but can delegate other rights to other
+accounts. We will like a secondary index on owner, so we can quickly
+display all NFTs controlled by your private key.
+
+`approvals: map[string][]Approval`: Every token may have approvals
+for different actions. Actions are named strings and for each one,
+there may be 0 or more accounts approved to execute it. We clearly
+cannot store a map in the kvstore due to determinism concerns, so
+I would suggest serializing it into arrays, such as 
+`[]{name: string, approvals: []Approval}`, and then sorting the arrays
+for consistency (top level by name string, each approval list sorted
+by the approved account).
+
+`payload: []byte/interface{}`: This is the species-specific part.
+I will not discuss more here, just mention a term for later reference.
+
+### Approval
+
+There are various ideas for fields as part of the approval info.
+The only one that is clear is account, which specifies who is approved.
+There are concepts of some metadata that would also accompany it.
+
+`account: Address`: The account which is approved for this action
+
+`timeout: int` ? : The approval may automatically expire at some block height (or timestamp?). This will not trigger a cron job, but usage after the timeout will just error (and trigger a cleanup maybe?)
+
+`immutible: boolean` ? : If ownership of the NFT is transfered,
+it may be normal for all Approvals to be revoked, so the new owner
+can re-issue them as needed. If I buy your domain name, I certainly
+don't expect your sysadmins to be able to set my DNS records.
+However, there are cases where the approval will persist beyond
+a transfer of ownership (like a lease of mineral rights to some land).
+The immutible flag could be used to renote the approval does not
+expire upon transfer of ownership.
```

Comment:
Q: How do we want to handle a combination of options. Immutible but timeout expired or immutible but count =0 ?

Reply:
A nitpick: should that be immutable instead?

Reply:
Yes, please add to the docs and push to this branch. With the spelling fix, but also with the clairifications that arise from your spike.

Or copy this doc into your spike and edit there if you intend to merge the spike later. In any case, this doc should make it into master when it is refined.